### ISSUE TITLE
Doesn't train actually.

### ISSUE THREAD

**Issue by chichivica** [meetshah1995/pytorch-semseg#31]
Many thanks for great opensource implementation of the semantic segmentation in pytorch ever! 
 
I'm trying to proceed through training 'segnet' model on 'pascal' dataset.
What I've done:
1) installed pytorch: 0.2.0_4 and python: 2.7.13
2) downloaded VOCtrainval_11-May-2012.tar from http://host.robots.ox.ac.uk/pascal/VOC/voc2012/#Development%20Kit
3) downloaded "Semantic Boundaries Dataset and Benchmark" from http://home.bharathh.info/pubs/codes/SBD/download.html
4) as stated in Readme, extracted and pointed to them in config.json file 
5) started training process
- started visdom server
 ```
python -m visdom.server
```
started training as:
```
python train.py --arch segnet --dataset pascal
```
training successfully started and going looks well:
![alt text](https://i.imgur.com/6PTZ8bJ.png "training process")

after completion, generated 100 **segnet_pascal_1_%2d.pkl files**

6) So, after that I'm trying to test newly trained model on the simple pictures:
```
python test.py --model_path segnet_pascal_1_99.pkl --dataset pascal --img_path 2007_000033.jpg --out_path result_33.jpg
```
![alt text](https://i.imgur.com/a1eqhkF.png "test model") 

But result is quite wrong:
![alt text](https://i.imgur.com/5qz4yIv.jpg "result 1") 

![alt text](https://i.imgur.com/IHot4td.jpg "result 2") 

for some reasons, output resolution differ and segmentation was not produced correctly.
 
Could you please give me some advises what I'm doing wrong?

![alt text](https://i.imgur.com/IAQfjxL.png "loss plot") 

Many thanks,
Ivan

**Comment by chichivica** [comment]
In addition, trained FCN8S, 
![alt text](https://i.imgur.com/4fienhu.jpg "fcn8s")

**Comment by meetshah1995** [comment]
Refer #20 , Try training fcn-8s without polynomial learning rate. Should work fine.

**Comment by chichivica** [comment]
@meetshah1995  trained SEGNET with disabled polynomial learning rate (commented out two lines):


[iter = len(trainloader)*epoch + i](https://github.com/meetshah1995/pytorch-semseg/blob/master/train.py#L60)
[poly_lr_scheduler(optimizer, args.l_rate, iter)](https://github.com/meetshah1995/pytorch-semseg/blob/master/train.py#L61)

![](https://i.imgur.com/eDScg3C.jpg "SEGNET segmentation results")

as well I had to change [here](https://github.com/meetshah1995/pytorch-semseg/blob/master/ptsemseg/loader/pascal_voc_loader.py#L106) like this:
```
rgb = np.zeros((temp.shape[1], temp.shape[2], 3))
```
It may be because I'm using python3, or because you [previously made a conversion](https://github.com/meetshah1995/pytorch-semseg/blob/master/test.py#L35) `NHWC -> NCWH`. I don't know. 

So, now I'm waiting util **fcn-8s** get trained to see results without polynomial learning rate.

**Comment by chichivica** [comment]
At least fcn-8s got trained. It took approximately 3 day on nvidia 1080. 
![](https://i.imgur.com/tC1ukMg.jpg)
![](https://i.imgur.com/PGsLjoZ.jpg)
![](https://i.imgur.com/x21Zfcw.jpg)
![](https://i.imgur.com/5f8t6aB.jpg)
![](https://i.imgur.com/rNnzj6l.jpg)
![](https://i.imgur.com/C4WsyJ1.jpg)

**Comment by XavierCHEN34** [comment]
hi， why does your results looks much worse than the  examples?

**Comment by meetshah1995** [comment]
PolyLR doesn't seem to work with FCN. FCN-8s in default settings work fine, with mIoU > 60

**Comment by HelloAlone** [comment]
hi, @chichivica if you want to train SegNet, you should set batch_size and l_rate, because of the BN layer. For example , you can set batch_size=16 and l_rate=1e-4.

**Comment by Galto2000** [comment]
I am getting the same (bad) results as @chichivica, with SegNet, even with the settings @HelloAlone suggested. My GTX1080Ti just could not handle a batch size of 16, so I used a batch size of 14 instead.

Also, the noise on my training curves increases significantly over time.  

![newplot 1](https://user-images.githubusercontent.com/19415965/42186462-cae83da0-7e1a-11e8-97b8-0443aa2cc427.png)

**Comment by xdsonglinliu** [comment]
@chichivica @Galto2000 
Thank you for your analysis.
I meet the same problem.
Have you finally trained the SegNet network successfully?
or Is there any other better SegNet implementation based on pytorch?
